cmd/errtrace/toolexec.go

```diff
@@ -0,0 +1,180 @@
+	}
+	// When "-toolexec" is used, the go cmd sets the package being compiled in the env.
+	if pkg := cmd.Getenv("TOOLEXEC_IMPORTPATH"); pkg != "" {
+		return cmd.toolExecRewrite(pkg, args), true
+	}
+
+	return -1, false
+}
+
+func (cmd *mainCmd) toolExecVersion(args []string) int {
+	tool := exec.Command(args[0], args[1:]...)
+	var stdout bytes.Buffer
+	tool.Stdout = &stdout
+	tool.Stderr = cmd.Stderr
+	tool.Env = os.Environ()
+	if err := tool.Run(); err != nil {
+		if exitErr, ok := err.(*exec.ExitError); ok {
+			return exitErr.ExitCode()
+		}
+
+		fmt.Fprintf(cmd.Stderr, "%v failed: %v", args[0], err)
+		return 1
+	}
+
+	// TODO: This version number should change whenever the rewriting changes.
+	fmt.Fprintf(cmd.Stdout, "%s-errtrace0\n", strings.TrimSpace(stdout.String()))
+	return 0
+}
+
+func (cmd *mainCmd) toolExecRewrite(pkg string, args []string) (exitCode int) {
+	// We only need to modify the arguments for "compile" calls which work with .go files.
+	if !isCompile(args[0]) {
+		return cmd.runOriginal(args)
+	}
+
+	// We only modify files that import errtrace, and have "errtrace.ToolExecInstrument"
+	// However, that requires parsing files to determine, which we want to avoid for stdlib
+	// so use a heuristic to detect stdlib packages -- whether the name contains "."".
+	if !strings.Contains(pkg, ".") {
+		return cmd.runOriginal(args)
+	}
+
+	exitCode, err := cmd.rewriteCompile(pkg, args)
+	if err != nil {
+		cmd.log.Print(err)
+		return 1
+	}
+
+	return exitCode
+}
+
+func (cmd *mainCmd) rewriteCompile(pkg string, args []string) (exitCode int, _ error) {
+	parsed := make(map[string]file)
+	var canRewrite, needRewrite bool
+	for _, arg := range args {
+		if !isGoFile(arg) {
+			continue
+		}
+
+		contents, err := os.ReadFile(arg)
+		if err != nil {
+			return -1, errtrace.Wrap(err)
+		}
+
+		f, err := cmd.parseFile(arg, contents)
+		if err != nil {
+			return -1, errtrace.Wrap(err)
+		}
+		parsed[arg] = f
+
+		// TODO: Support an "unsafe" mode to rewrite packages without errtrace imports.
+		if f.importsErrtrace {
+			canRewrite = true
+		}
+		if len(f.inserts) > 0 {
+			needRewrite = true
+		}
+	}
+
+	if !canRewrite || !needRewrite {
+		return cmd.runOriginal(args), nil
+	}
+
+	// Use a temporary directory per-package that is rewritten.
+	tempDir, err := os.MkdirTemp("", filepath.Base(pkg))
+	if err != nil {
+		return -1, errtrace.Wrap(err)
+	}
+
+	newArgs := make([]string, 0, len(args))
+	for _, arg := range args {
+		if !isGoFile(arg) {
+			newArgs = append(newArgs, arg)
+			continue
+		}
+
+		f, ok := parsed[arg]
+		if !ok || len(f.inserts) == 0 {
+			newArgs = append(newArgs, arg)
+			continue
+		}
+
+		// Add a //line directive so the original filepath is used in errors and panics.
+		out := bytes.NewBuffer(nil)
+		_, _ = fmt.Fprintf(out, "//line %v:1\n", arg)
+
+		if err := cmd.rewriteFile(f, out); err != nil {
+			return -1, errtrace.Wrap(err)
+		}
+
+		newFile := filepath.Join(tempDir, filepath.Base(arg))
```

Comment:
We'll want to use `os.CreateTemp(tempDir, ... "*.go")` here. I don't think it's guaranteed that the base name will be unique. For example, if invoked by Bazel, your files can be coming from different directories. AFAIK, they can use the same base names.

Fine to leave it as a TODO for later, or for when it becomes a problem.

Reply:
Ah bazel fun, I'll leave a TODO for now, would prefer to keep it simple for now.